Provide a way to crash a WebKit XPC service for testing

msaboff opened the pull request:
#### 3f65989a37af5b77e9ffb4e4918e793ffe685135
<pre>
Provide a way to crash a WebKit XPC service for testing
<a href="https://bugs.webkit.org/show_bug.cgi?id=246782">https://bugs.webkit.org/show_bug.cgi?id=246782</a>
rdar://99339140

Reviewed by Chris Dumez.

Added a new WebContent.Crashy XPC process specifically to crash.  Safari, when running on internal builds will launch
a WebContent.Crashy depending on the canary base state.  A crashy WebContent process is launch immediately after
a WebContent for prewarming.  The UIProcess code will reap the crashed process without any impact on Safari.
This crashing process is used for internal testing.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/Configurations/WebContentService.Crashy.xcconfig: Added.
* Source/WebKit/Configurations/WebContentService.xcconfig:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.mm:
(WebKit::XPCServiceInitializerDelegate::getExtraInitializationData):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(isInternalBuild):
(canaryInBaseState):
(WebKit::determineIfWeShouldCrashWhenCreatingWebProcess):
(WebKit::WebProcessPool::platformInitialize):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createNewWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createForWebContentCrashy):
(WebKit::WebProcessProxy::getLaunchOptions):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::isCrashyProcess):
(WebKit::WebProcessProxy::setIsCrashyProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::deliberateCrashForTesting):
(WebKit::WebProcess::platformInitializeProcess):

Canonical link: <a href="https://commits.webkit.org/255828@main">https://commits.webkit.org/255828@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ea80ff2e646abcf5a9a8d47079f039eba4bbf328

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93707 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2902 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24318 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103351 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/163671 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2915 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31160 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86046 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99395 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99368 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2073 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80145 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29102 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/97292 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83997 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72054 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebView:/webkit/WebKitWebView/title-change (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/84967 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37551 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17565 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80092 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35401 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/18825 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27639 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4029 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39278 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41370 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/82733 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41213 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38056 "Passed tests") | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18706 "Found 1 new JSC stress test failure: stress/call-apply-exponential-bytecode-size.js.mini-mode (failure)") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -447,6 +447,10 @@ class WebProcessPool final
 #endif
 
+#if ENABLE(WEBCONTENT_CRASH_TESTING)
+    bool crashWhenCreatingWebProcess() const { return m_crashWhenCreatingWebProcess; }
```
I think both the variable and function name should start with "should" for clarity. The current naming is a little bit confusing in my opinion.

msaboff replied:
I'll make this change.

cdumez commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -207,6 +207,16 @@ Ref<WebProcessProxy> WebProcessProxy::createForRemoteWorkers(RemoteWorkerType wo
+Ref<WebProcessProxy> WebProcessProxy::createForWebContentCrashy(WebProcessPool& processPool)
+{
+    auto proxy = adoptRef(*new WebProcessProxy(processPool, nullptr, IsPrewarmed::No, CrossOriginMode::Shared, LockdownMode::Disabled));
+    proxy->setCrashyProcess();
```
I think we usually would call this `setIsCrashyProcess()`, with the getter being `isCrashyProcess()`

msaboff replied:
I'll make this change.

cdumez reviewed (approved):
LGTM

msaboff committed:
Provide a way to crash a WebKit XPC service for testing
https://bugs.webkit.org/show_bug.cgi?id=246782
rdar://99339140

Reviewed by Chris Dumez.

Added a new WebContent.Crashy XPC process specifically to crash.  Safari, when running on internal builds will launch
a WebContent.Crashy depending on the canary base state.  A crashy WebContent process is launch immediately after
a WebContent for prewarming.  The UIProcess code will reap the crashed process without any impact on Safari.
This crashing process is used for internal testing.

* Source/WTF/wtf/PlatformEnableCocoa.h:
* Source/WebKit/Configurations/WebContentService.Crashy.xcconfig: Added.
* Source/WebKit/Configurations/WebContentService.xcconfig:
* Source/WebKit/Scripts/process-entitlements.sh:
* Source/WebKit/Shared/EntryPointUtilities/Cocoa/XPCService/XPCServiceEntryPoint.mm:
(WebKit::XPCServiceInitializerDelegate::getExtraInitializationData):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(isInternalBuild):
(canaryInBaseState):
(WebKit::determineIfWeShouldCrashWhenCreatingWebProcess):
(WebKit::WebProcessPool::platformInitialize):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::createNewWebProcess):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::createForWebContentCrashy):
(WebKit::WebProcessProxy::getLaunchOptions):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::isCrashyProcess):
(WebKit::WebProcessProxy::setIsCrashyProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/cocoa/WebProcessCocoa.mm:
(WebKit::deliberateCrashForTesting):
(WebKit::WebProcess::platformInitializeProcess):

Canonical link: https://commits.webkit.org/255828@main
